### PR TITLE
CHE-13112: Respect workspaces based on devfile in Che Theia IDE UI

### PR DESCRIPTION
**Pull request by mmorhun**
Signed-off-by: Mykola Morhun <mmorhun@redhat.com>

### What does this PR do?
Adds functionality to support devfile configuration of workspace on Theia UI. So now it supports both: devfile and workspace config.
The changes is mostly about projects sync between Theia and devfile projects section.
Also adds small changed to workspace panel to be able to show appropriate info from devfile. But with current state we cannot get information about volumes, so in case of devfile config the information about volumes is missing for now. See https://github.com/eclipse/che/issues/13290.

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/13112

### CODE REVIEW

**Comment by olexii4** [comment]
https://github.com/eclipse/che-theia/blob/374f5b2dfeabf540d6377e51c905541e9e0e033a/plugins/containers-plugin/src/containers-tree-data-provider.ts#L109

command: { id: 'task:run', arguments: ['che', commandName] }  -> command: { id: 'task:run', arguments: ['che', commandName, container.name] }

WDYT?

File: plugins/containers-plugin/src/containers-service.ts [hunk 1]
```diff
@@ -47,23 +52,46 @@ export class ContainersService {
...
+                    // should never happen
+                    throw new Error('Invalid workspace structure');
+                }
             }
-            if (workspace.runtime!.commands) {
+            if (runtime.commands) {
                 container.commands = [];
-                workspace.runtime.commands.forEach(command => {
+                runtime.commands.forEach(command => {
                     if (command.attributes && command.attributes.machineName && command.attributes.machineName !== name) {
```
Comment (AndrienkoAleksandr) [inline plugins/containers-plugin/src/containers-service.ts]: So, if command without attributes  at all, we push this command to the container? Maybe for such commands we should apply attribute with machineName manually and than push to the container command?

File: plugins/factory-plugin/src/projects.ts [hunk 2]
```diff
@@ -7,13 +7,88 @@
...
+    projectPath: string,
+    projectGitLocation: string,
+    projectGitRemoteBranch: string
+): cheApi.workspace.devfile.Project[] {
+
+    if (!projects) {
+        projects = [];
+    }
+
+    for (let i = 0; i < projects.length; i++) {
```
Comment (AndrienkoAleksandr) [inline plugins/factory-plugin/src/projects.ts]: Maybe `for...of` will be better? But up to you.

File: plugins/containers-plugin/src/containers-service.ts (hunk 1 above)
Comment (mmorhun) [reply plugins/containers-plugin/src/containers-service.ts]: I'll handle it in https://github.com/eclipse/che-theia/issues/222

File: plugins/factory-plugin/src/workspace-projects-manager.ts [hunk 3]
```diff
@@ -66,70 +70,150 @@ export class WorkspaceProjectsManager {
     async startSyncWorkspaceProjects() {
         const gitConfigPattern = '**/.git/{HEAD,config}';
         const gitConfigWatcher = theia.workspace.createFileSystemWatcher(gitConfigPattern);
-        gitConfigWatcher.onDidCreate(uri => this.updateOrCreateGitProjectInWorkspace(git.getGitRootFolder(uri.path)));
-        gitConfigWatcher.onDidChange(uri => this.updateOrCreateGitProjectInWorkspace(git.getGitRootFolder(uri.path)));
-        gitConfigWatcher.onDidDelete(uri => this.deleteGitProjectInWorkspace(git.getGitRootFolder(uri.path)));
+        gitConfigWatcher.onDidCreate(uri => this.updateOrCreateProjectInWorkspace(git.getGitRootFolder(uri.path)));
```
Comment (AndrienkoAleksandr) [inline plugins/factory-plugin/src/workspace-projects-manager.ts]: onDidCreate, onDidChange, onDidDelete returns Disposable, push them to the plugin context please.

Comment (mmorhun) [reply plugins/factory-plugin/src/workspace-projects-manager.ts]: thanks!

Comment (mmorhun) [reply plugins/factory-plugin/src/workspace-projects-manager.ts]: They will be disposed when the watcher gets disposed

**Pull request merged**